File: ga_python/GA.py
"""This module implements Genetic Algorithms."""
import numpy as np
from multiprocessing import Pool


def rand_population_uniform(m, a, b):
    """
    Creates a starting population of size `m` using a uniform random
    distribution, where `a` is the lower bound, and `b` is the upper bound
    for the design variables.
    """
    a = np.array(a)
    b = np.array(b)
    d = len(a)
    return [list(np.multiply(a + np.random.rand(d), (b-a))) for i in range(m)]


def TruncatedSelection(pop, tr):
    """Performs truncation selection for top `tr` candidates."""
    p = np.argsort(pop)
    return [p[np.random.randint(tr, size=(1, 2)).ravel()] for i in p]


def SinglePointCrossover(a, b):
    """Performs a single point crossover for two parents, `a` and `b`."""
    a = list(a)
    b = list(b)
    n = len(a)
    i = np.random.randint(n+1)
    return a[:i]+b[i:]


def GaussianMutation(child, sigma):
    """
    Performs a Gaussian mutation for a child candidate, with scaling factor
    `sigma`.
    """
    if np.random.rand() < sigma:
        new_child = child + np.random.rand(len(child))*sigma
    else:
        new_child = child
    return new_child


def GeneticAlgorithm(f, population, k_max=10, sigma=0.1, trunc=10,
                     processes=None):
    """
    Performs a Genetic Algorithm optimization with optional parallelization.

    Args:
        f (Function): Objective function.
        population (list of lists): A list of design variable combinations
        representing an initial population.
        k_max (int): Maximum iterations.
        sigma (float): Scaling factor for Gaussian mutation.
        trunc (int): Cutoff for truncation selection.
        processes (int): Number of processes in the `Pool`.

    Returns:
        Array: An Array representing the minimizers for the given objective
        function.
    """
    if processes:
        p = Pool(processes=processes)

        for _ in range(k_max):
            results = []
            for x in population:
                res = p.apply_async(f, [x])
                results.append(res)

            f_pop = [r.get() for r in results]

            parents = np.array(TruncatedSelection(f_pop, trunc))
            children = [
                SinglePointCrossover(population[p[0]],
                                     population[p[1]]) for p in parents]

            population = [GaussianMutation(child, sigma) for child in children]

    else:
        for k in range(k_max):
            f_pop = list([f(np.array(x)) for x in population])
            parents = np.array(TruncatedSelection(f_pop, trunc))

            children = [
                SinglePointCrossover(population[p[0]],
                                     population[p[1]]) for p in parents]

            population = [GaussianMutation(child, sigma) for child in children]

    f_pop = list([f(np.array(x)) for x in population])
    return population[np.argmin(f_pop)]

File: ga_julia/GA.jl
"""
Creates a starting population of size `m` using a uniform random distribution,
where `a` is the lower bound, and `b` is the upper bound for the design
variables.
"""
function rand_population_uniform(m, a, b)
    d = length(a)
    return [a+rand(d).*(b-a) for i in 1:m]
end

abstract type SelectionMethod end
struct TruncationSelection <: SelectionMethod
    k # top k to keep
end
function select(t::TruncationSelection, y)
    p = sortperm(y)
    return [p[rand(1:t.k, 2)] for i in y]
end


abstract type CrossoverMethod end
struct SinglePointCrossover <: CrossoverMethod end
function crossover(::SinglePointCrossover, a, b)
    i = rand(1:length(a))
    return vcat(a[1:i], b[i+1:end])
end

abstract type MutationMethod end
struct GaussianMutation <: MutationMethod
    σ
end
function mutate(M::GaussianMutation, child)
    return child + randn(length(child))*M.σ
end


"""
Performs a Genetic Algorithm optimization with optional parallelization.

  Args:
    f (Function): Objective function.
    population (Array[]): An Array of design variable combinations
      representing an initial population.
    k_max (Integer): Maximum iterations
    S (SelectionMethod): The selection method to be used.
    C (CrossoverMethod): The crossover method to be used.
    M (MutationMethod): The mutation method to be used.
    parallel (Boolean, optional): Optionally use multithreading.

  Returns:
    Array: An Array representing the minimizers for the given objective
      function.
"""
function genetic_algorithm(f, population, k_max, S, C, M; parallel=false)
    n = length(population)
    for k in 1 : k_max
        if parallel
            # Perform function evaluations in parallel
            f_pop = zeros(n)
            Threads.@threads for i=1:n
               f_pop[i] = f(population[i]) 
            end
        else
            f_pop = f.(population)
        end
        # proceed as normal for selection, crossover, mutation
        parents = select(S, f_pop)
        children = [crossover(C, population[p[1]], population[p[2]]) for p in parents]
        population .= mutate.(Ref(M), children)
    end

    if parallel
        # Perform function evaluations in parallel
        f_pop = zeros(n)
        Threads.@threads for i=1:n
            f_pop[i] = f(population[i]) 
        end
    else
        f_pop = f.(population)
    end
    population[argmin(f_pop)]
end
